src/twisted/web/client.py

```diff
@@ -955,6 +956,65 @@ def creatorForNetloc(self, hostname, port):
                           WebClientContextFactory.__name__)
 
 
+@implementer(IPolicyForHTTPS)
+class HostnameCachingHTTPSPolicy(object):
+    """
+    SSL connection creator for web clients that caches the last cacheSize
+    L{client connection creators <twisted.internet.interfaces.
+    IOpenSSLClientConnectionCreator>} for reuse in subsequent requests.
+
+    @ivar _trustRoot: See C{trustRoot} parameter of L{__init__}.
+
+    @ivar _cache: A cache associating hostnames to their
+        L{_HostnameCacheEntry}.
+    @type _cache: L{dict}
+
+    @ivar _cacheSize: See C{cacheSize} parameter of L{__init__}.
+    """
+
+    def __init__(self, trustRoot=None, cacheSize=20):
+        """
+        @param trustRoot: See the C{trustRoot} parameter of
+            L{optionsForClientTLS}.
+        @type trustRoot: L{IOpenSSLTrustRoot}
+
+        @param cacheSize: The maximum size of the hostname cache.
+        @type cacheSize: L{int}
+        """
+        self._trustRoot = trustRoot
+        self._cache = collections.OrderedDict()
+        self._cacheSize = cacheSize
+
+
+    def creatorForNetloc(self, hostname, port):
+        """
+        Create a L{client connection creator
+        <twisted.internet.interfaces.IOpenSSLClientConnectionCreator>} for a
+        given network location and cache it for future use.
+
+        @param hostname: The hostname part of the URI.
+        @type hostname: L{bytes}
+
+        @param port: The port part of the URI.
+        @type port: L{int}
+
+        @return: a connection creator with appropriate verification
+            restrictions set
+        @rtype: L{client connection creator
+            <twisted.internet.interfaces.IOpenSSLClientConnectionCreator>}
+        """
+        host = hostname.decode("ascii")
+        try:
+            creator = self._cache.pop(host)
+        except KeyError:
+            creator = optionsForClientTLS(host, trustRoot=self._trustRoot)
```

Comment:
Instead of duplicating the logic in `BrowserLikePolicyForHTTPS` it would be nice to make this implementation a wrapper around an arbitrary other implementation.  This will also simplify testing.

Reply:
Seems reasonable and not too difficult.  I'll make the changes and push here to make sure I'm on the right track (I find code makes any subsequent discussions easier) before putting it back into review.

Reply:
@exarkun Is this along the lines of what you had in mind?
https://github.com/jlitzingerdev/twisted/blob/d48f24ec36683c546486d5878216427a25c654a3/src/twisted/web/client.py#L1010-1021